Add ModelarDB compression evaluation script

skejserjensen opened the pull request:
This PR adds a script that evaluates how well ModelarDB compresses real-life data sets for a set of error bounds.

CGodiksen commented on Evaluate-ModelarDB-Compression/main.py:
```diff
@@ -0,0 +1,281 @@
+""" Script for evaluating ModelarDB with different data sets. """
```
This comment should not have a leading and trailing space. It also does not seem like the other utility files have a module doc comment so this could also be removed.

skejserjensen replied:
The formatting was copied from [Evaluate-ModelarDB-Changes/main.py](https://github.com/ModelarData/Utilities/blob/main/Evaluate-ModelarDB-Changes/main.py) so I have removed the spaces in both `main.py` files.

CGodiksen commented on Evaluate-ModelarDB-Compression/main.py:
```diff
@@ -0,0 +1,281 @@
+""" Script for evaluating ModelarDB with different data sets. """
+
+
```
Maybe run a formatting tool on the file. I think there should only be a single newline between the module doc comment and imports.

skejserjensen replied:
The file was already formatted using [black](https://github.com/psf/black), so I manually removed one of the empty lines from `Evaluate-ModelarDB-Compression/main.py` and `Evaluate-ModelarDB-Changes/main.py`.

CGodiksen commented on Evaluate-ModelarDB-Compression/main.py:
```diff
@@ -0,0 +1,281 @@
+import numpy
+from pyarrow import parquet
+from pyarrow import flight
+
```
Maybe run a formatting tool on the file. I think there should only be a single newline between imports and module constants.

CGodiksen replied:
It seems that `Evaluate-ModelarDB-Changes/main.py` might have the same formatting issues so maybe also fix it there.

skejserjensen replied:
The file was already formatted using [black](https://github.com/psf/black), so I manually removed one of the empty lines from `Evaluate-ModelarDB-Compression/main.py` and `Evaluate-ModelarDB-Changes/main.py`.

CGodiksen commented on Evaluate-ModelarDB-Compression/main.py:
```diff
@@ -0,0 +1,281 @@
+    subprocess.run(["git", "clone", url], stdout=STDOUT, stderr=STDERR)
+
+
+def cargo_build_release(modelardb_folder):
```
This might be overkill for a utility repository but it seems that there is quite a bit of duplicated code between this file and `Evaluate-ModelarDB-Changes/main.py`. Maybe create a utility file with the common functionality.

skejserjensen replied:
I also came to the conclusion that it is overkill for now due to the number of and size of scripts we currently have. However, I do think it will become a problem if we continue adding more scripts and extend the existing scripts. Thus, I have created a low priority task to move shared functionality into a separate shared Python module.

CGodiksen commented on Evaluate-ModelarDB-Compression/main.py:
```diff
@@ -0,0 +1,281 @@
+        )
+    )
+    print("- Error Ceil Histogram:", end="")
+    for ceiled_error_cound, count in enumerate(ceiled_actual_errors):
```
Maybe `ceiled_error_count` instead.

skejserjensen replied:
Fixed.

CGodiksen commented on Evaluate-ModelarDB-Compression/main.py:
```diff
@@ -0,0 +1,281 @@
+
+
+def retrieve_ingested_column(flight_client, column_name, timestamp_column):
+    ticket = flight.Ticket(
```
Maybe use an [f-string](https://peps.python.org/pep-0498/) instead to make it a little more readable.

skejserjensen replied:
Fixed.

CGodiksen commented on Evaluate-ModelarDB-Compression/main.py:
```diff
@@ -0,0 +1,281 @@
+            return
+
+    # Compute and print the final result.
+    print("- Real Values: {}".format(len(real_column)))
```
Maybe use [f-strings](https://peps.python.org/pep-0498/) instead to make it a little more readable.

skejserjensen replied:
Fixed.

CGodiksen commented on Evaluate-ModelarDB-Compression/main.py:
```diff
@@ -0,0 +1,281 @@
+if __name__ == "__main__":
+    # Ensure the necessary arguments are provided.
+    if len(sys.argv) < 2:
+        print("usage: " + sys.argv[0] + "test_data.parquet error_bound*")
```
Maybe use an [f-string](https://peps.python.org/pep-0498/) instead to make it a little more readable. This also applies to all other places in `main` where either `.format()` is used or `"string" + arg` is used.

skejserjensen replied:
Fixed.

CGodiksen commented on Evaluate-ModelarDB-Compression/main.py:
```diff
@@ -195,20 +185,20 @@ def send_sigint_to_process(process):
     # The script assumes it runs on Linux.
     if sys.platform != "linux":
-        print("ERROR: " + sys.argv[0] + " only supports Linux")
+        print(f"ERROR: sys.argv[0] only supports Linux")
```
Missing `{}` around `sys.argv[0]`.

skejserjensen replied:
Fixed.

chrthomsen commented on Evaluate-ModelarDB-Compression/main.py:
```diff
@@ -0,0 +1,271 @@
+
+    # Compute and print the final result.
+    print(f"- Real Values: {len(real_column)}")
+    print(f"- Compressed Values: {len(compressed_column)}")
```
it could also be checked in the beginning that these have the same length. Shouldn't there be an exception if they have different lengths?

skejserjensen replied:
I contemplated adding such an exception, but decided against it to allow the script to be used where the compressed column contains the `n` first values of the real column. However, to make this more clear, the script now print a warning if the compressed column is shorter, and an error if the real column is shorter.

chrthomsen commented on Evaluate-ModelarDB-Compression/main.py:
```diff
@@ -0,0 +1,271 @@
+    print(f"- Real Values: {len(real_column)}")
+    print(f"- Compressed Values: {len(compressed_column)}")
+    print(f"- Without Error: {100 * (equal_values / compared_values)}%")
+    print(f"- Average Error: {100 * abs(sum_difference / sum_real_values)}%")
```
wouldn't the average *absolute* error be relevant? In an extreme case, there could be errors for each data point such that the errors are of the form: +0.1, -0.1, +0.1, -0.1, ... (assuming they are within the error bound). Then the average is 0.

skejserjensen replied:
I have added `sum_absolute_difference / len(compressed_column)`, was that what you had in mind?

chrthomsen replied:
I was more thinking of the average of the `actual_error` values, i.e., the MAPE. Not sure if the average of the absolute errors is useful

skejserjensen replied:
Fixed.

chrthomsen commented on README.md:
```diff
@@ -17,6 +17,13 @@ Utilities for simplifying development and testing:
+  [ModelarDB](https://github.com/ModelarData/ModelarDB-RS) compresses a data set
+  stored in Apache Parquet files with the same schema for a set of error bounds.
+  For each error bound the script automatically ingests the data set and
+  computes metrics that indicates how well ModelarDB compressed the data set.
```
metrics that indicate

skejserjensen replied:
Fixed.

skejserjensen commented on Evaluate-ModelarDB-Compression/main.py:
```diff
@@ -0,0 +1,271 @@
+    # Compute metrics.
+    for real_value, compressed_value in zip(real_column, compressed_column):
+        if real_value == compressed_value:
+            equal_values += 1.0
```
Fixed.

skejserjensen commented on Evaluate-ModelarDB-Compression/main.py:
```diff
@@ -0,0 +1,271 @@
+    max_actual_error_real_value = 0.0
+    max_actual_error_compressed_value = 0.0
+
+    ceiled_error_counts = math.ceil(error_bound + 2) * [0]
```
Fixed.

skejserjensen commented on Evaluate-ModelarDB-Compression/main.py:
```diff
@@ -0,0 +1,271 @@
+
+        try:
+            ceiled_error_counts[math.ceil(actual_error)] += 1
+        except OverflowError:
```
`math.ceil()` raises errors if `actual_error` is a float specific value. I have added a comment explain this.

skejserjensen commented:
I have updated the script based on the comments from @chrthomsen and fixed `sum_difference` and `sum_real_values` not accumulating the absolute values.
